content/en/blog/_posts/2023-09-15-Bootstrap-an-Air-Gapped-Cluster-with-Kubeadm/index.md

```diff
@@ -0,0 +1,665 @@
+
+```bash
+UARCH=$(uname -m)
+# Set architecture variables
+
+if [["$UARCH" == "arm64" || "$UARCH" == "aarch64"]]; then
+
+    ARCH="aarch64"
+    K8s_ARCH="arm64"
+
+else
+
+    ARCH="x86_64"
+    K8s_ARCH="amd64"
+
+fi
+
+# Set environmental vartiables
+CNI_PLUGINS_VERSION="v1.3.0"
+CRICTL_VERSION="v1.27.0"
+KUBE_RELEASE="v1.27.3"
+RELEASE_VERSION="v0.15.1"
+K9S_VERSION="v0.27.4"
+
+cd ~/tmp/
+
+dnf -y install ./*.rpm
+```
+
+**Next, install the CNI plugins and crictl:**
+
+
+```bash
+mkdir -p /opt/cni/bin
+tar -C /opt/cni/bin -xz -f "cni-plugins-linux-${K8s_ARCH}-v1.3.0.tgz"
+tar -C /usr/local/bin-xz -f "crictl-v1.27.0-linux-${K8s_ARCH}.tar.gz"
+```
+
+**Make kubeadm, kubelet and kubectl executable and move them from the /tmp directory to /usr/local/bin:**
+```bash
+chmod +x kubeadm kubelet kubectl
+mv kubeadm kubelet kubectl /usr/local/bin
+```
+**Define an override for the systemd kubelet service file and move it to the proper location.**
+
+```bash
+mkdir -p /etc/systemd/system/kubelet.service.d
+
+sed "s:/usr/bin:/usr/local/bin:g" 10-kubeadm.conf > /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+```
+
+**The CRI plugin for containerd is disabled by default; enable it:**
+```bash
+sed -i 's/^disabled_plugins = \["cri"\]/#&/' /etc/containerd/config.toml
+```
+**Put the /etc/docker/daemon.json file in place:**
+```bash
+echo '{
+
+"exec-opts": ["native.cgroupdriver=systemd"],
+
+"insecure-registries" : ["localhost:5000"],
+
+"allow-nondistributable-artifacts": ["localhost:5000"],
+
+"log-driver": "json-file",
+
+"log-opts": {
+
+"max-size": "100m"
+
+},
+
+"group": "rnd",
+
+"storage-driver": "overlay2",
+
+"storage-opts": [
+
+"overlay2.override_kernel_check=true"
+
+]
+
+}' > /etc/docker/daemon.json
+```
+Two important lines to highlight in the daemon.json file. The insecure-registries line means that the registry in brackets does not support TLS. Even inside an air gapped environment, this isn't a good practice but is fine for the purposes of this lab. The allow-nondistributable-artifacts line tells Docker to permit pushing nondistributable artifacts to this registry. Docker by default does not push these layers to avoid potential issues around licensing or distribution rights. A good example of this is the Windows base container image. This line will allow layers that Docker marks as "foreign" to be pushed to the registry. While not a big deal for this article, that line could be required for some air gapped environments. All layers have to exist locally since nothing inside the air gapped environment can reach out to a public container image registry to get what it needs.
```

Comment:
```suggestion
Two important items to highlight in the Docker `daemon.json` configuration file.
The `insecure-registries` setting means that the registry in brackets does not support TLS.
Even inside an air gapped environment, this isn't a good practice - but is fine for the purposes
of this article and the lab environment I'm describing.
The `allow-nondistributable-artifacts` line tells Docker to permit pushing nondistributable
artifacts to this registry. Docker by default does not push these layers to avoid potential issues
around licensing or distribution rights. A good example of this is the Windows base container
image. This configuration will allow layers that Docker marks as _foreign_ to be pushed to the
registry on the internal side of the air gap. While not a big deal for this article, that line could
be required for some air gapped environments. All layers have to exist locally since nothing
inside the air gapped environment can reach out to a public container image registry to get
what it needs.
```

Reply:
@sftim , added this suggestion manually (Github was going to break this chunk up into multiple lines if I committed it directly from the comment 🤷‍♂️ ).